### ISSUE TITLE
SWIG incorrectly decides a class is abstract

### ISSUE THREAD

**Issue by ojwb** [swig/swig#2333]
This was https://sourceforge.net/p/swig/bugs/1321/ originally reported in 2013 against SWIG 2.0.9 but still present in current SWIG.

Here's a reproducer I adapted from the incomplete one provided in the original ticket:

```
$ cat test.i
%module "inherit"

class a {
public :
virtual void test(void *p = 0) = 0;
};

class b : public a {
public:
void test(void *p) {}
b() {}
};
$ swig -c++ -python test.i
$ grep wrap_new_ test_wrap.cxx
$
```

C++ class `a` is abstract, but class `b` isn't.  However SWIG fails to understand that `void test(void *p)` implements the virtual method and also treats `b` as abstract and doesn't wrap its constructor.  There should be a `wrap_new_b` function found by the `grep` (and there is if `= 0` is added to the `p` parameter in class `b`).

This also affect PHP - I doubt it's specific to the target language at all.

**Comment by Gonzalo-Mier** [comment]
Hi @ojwb, 
First of all, thanks for this amazing package!!

About this issue, I don't think this should be corrected, as it promotes good C++ practices and reduce future errors (I don't actually know why this compiles on C++ in the first place). Please, tell me if I'm wrong.

```
class a {
public :
virtual void test(void *p = 0) = 0;
};
```
may produce two functions interfaces:
- `a::test(void *p)` virtual and not defined (not created)
- `a::test()` virtual and not defined (not created)

Then **b** inherits:
```
class b : public a {
public:
void test(void *p) {}
b() {}
};
``` 
which may produce four possible functions interfaces:
- **`b::test(void *p)`** non-virtual and defined.
- **`a::test(void *p)`** virtual and not defined, inherited from `a::test(void *p)`, and currently non-accessible, shadowed by non-virtual `b::test(void *p)`
-  `b()` defined

So **b** has some functions which are not defined.  `b::test()` is not accessible despite is inherited from `a::test()` because it's shadowed from `b::test(void *p)`

The correct example should be:
``` 
class a {
public :
virtual void test(void *p = 0) = 0;
};

class b : public a {
public:
void test(void *p = 0) override {}
b() {}
};
``` 
which definitively works!

**Comment by ojwb** [comment]
I agree this is rather a corner case, and arguably a situation that's better avoided, but SWIG aims to support standard C++, and standard C++ supports this so I think it's clearly a bug in SWIG. 

Also, the underlying issue here seems to be that SWIG's model of how this works isn't the same as what the standard defines, and that is liable to manifest in other ways.

It's probably not a high priority to address (indeed nobody has for ~10 years since the original report, though for much of that time the report was buried in the old tracker which I suspect I'm the only one to still look at).

**Comment by ojwb** [comment]
Reproducible with de7fbae696094b4c03a68d9821e93d887a848287.

**Comment by KrisThielemans** [comment]
I discovered we have this problem in our code base, see the STIR issue link above. At least this bug report allows me to implement a work-around. Thanks @ojwb
